### ISSUE TITLE
Cleanup not working: 'Can't perform a React state update on an unmounted component.'

### ISSUE THREAD

**Issue by ptisdel** [use-global-hook/use-global-hook#40]
I'm getting the following warning:

> Can't perform a React state update on an unmounted component. This is a no-op, but it indicates a memory leak in your application. To fix, cancel all subscriptions and asynchronous tasks in a useEffect cleanup function.

Here's my (simplified) code.

**App.js**
```
// if a playerId exists in global state, render PlayerPage, otherwise LoginPage.

export const App = () => {
  const [globalState, globalActions] = useGlobal();

  const { playerId } = globalState;

  const renderView = () => {
     if (playerId) return <PlayerPage/>;
     return <LoginPage/>;
  };

   return (
     { renderView() }
    );
};
```

**PlayerPage.js**
```
// display playerId from globalState

const PlayerPage = () => {
  const [globalState, globalActions] = useGlobal();
  const { playerId } = globalState;

  return <div>{ playerId }</div>;
};
```

This works, but if `playerId` in GlobalState is changed by GlobalAction to null, I get the warning above when it tries to perform a state update on the unmounted PlayerPage. Since useGlobalHook handles cleanup on its own, I'm not sure why I'm getting this warning.

**Comment by xiaomocqk** [comment]
I have the same issue

**Comment by ngoductaka** [comment]
in node_module/use-global-hook/index.js:
```
function setState(store, newState, afterUpdateCallback) {
  // store.state = { ...store.state, ...newState }; // replate its
  store.state = { ...newState };
  store.listeners.forEach((listener) => {
    listener.run(store.state);
  });
  afterUpdateCallback && afterUpdateCallback();
}

```

**Comment by sanjups7** [comment]
I also have this issue, in my case this is happening when redirected to a new page after a async action is done.
`
const isMountedRef = React.useRef(null);
const [globalState, globalActions] = useGlobal();
const { isLogged } = globalState.appContext;
const { login } = globalActions.appContext;
React.useEffect(() => {
	isMountedRef.current = true;
	if (isLogged) {
		if (isMountedRef.current) {
			history.push("/app");
		}
	}
	return () => (isMountedRef.current = false);
}, [isLogged]);
const handleSubmit = (formData) => {
	login(formData);
};
`

**Comment by theceremony** [comment]
yeah I'm getting the same issue, is there a way to unsubscribe from the listeners in the cleanup on useEffect, seems like that could be a solution?

**Comment by jamesacklin** [comment]
This is also an issue for me.

**Comment by ElRatonDeFuego** [comment]
I've been bitten by this as well, know where the problem, and have a fix. The problem is here : https://github.com/andregardi/use-global-hook/blob/master/src/core/newListenerEffect.js#L5-L15

The state updater functions (either `originalHook` or the arrow function) should act as no-ops if the component is in the process of being unmounted. This could be done with a flag set to `true` in the cleanup function. I think I'm going to write a proper PR for this, here is what the new `newListenerEffect.js` could look like:

```
import { cleanUpListener } from "./cleanUpListener";

export const newListenerEffect = (store, mapState, originalHook) => () => {
  let isComponentBeingUnmounted = false;
  const newListener = { oldState: {} };
  newListener.run = mapState
    ? (newState) => {
        if (isComponentBeingUnmounted) {
          return;
        }
        const mappedState = mapState(newState);
        if (mappedState !== newListener.oldState) {
          newListener.oldState = mappedState;
          originalHook(mappedState);
        }
      }
    : () => {
        if (!isComponentBeingUnmounted) {
          originalHook();
        }
      };
  store.listeners.push(newListener);
  return () => {
    isComponentBeingUnmounted = true;
    cleanUpListener(store, newListener)();
  };
};

```

**Comment by ElRatonDeFuego** [comment]
Hello, I've fixed this (as well as another issue) in a PR : https://github.com/andregardi/use-global-hook/pull/59

The problem was that, wen a component is being unmounted, it should not attempt to update its state.

If the body of useEffect functions may asynchronously attempt to update the state (as was the case here), they need to *not* call setState once the useEffect cleanup function has been called. This can be done by setting a flag, as per the PR.

You can test this fix with: `npm i --save ElRatonDeFuego/use-global-hook#0.2.1-1`
(for TypeScript, it still works with the types from `npm i --save-dev @types/use-global-hook`)

**Comment by Twansparant** [comment]
Same issue for me and can confirm that @ElRatonDeFuego **PR** fixed it for me.
Thanks!

**Comment by ElRatonDeFuego** [comment]
I've added another fix to my PR : https://github.com/ElRatonDeFuego/use-global-hook/commit/d2fbebfa9ee16931b19e09914457e3b19056c97d

You can test with `npm i --save ElRatonDeFuego/use-global-hook#0.2.1-2`

**Comment by ElRatonDeFuego** [comment]
I've added yet another fix to my PR to prevent unnecessary re-renders on the first update: https://github.com/ElRatonDeFuego/use-global-hook/commit/ccbe16756de0a8634a840cdc12cbe7a3aae70bb3

You can test with `npm i --save ElRatonDeFuego/use-global-hook#0.2.1-3`
